AK/Format.h

```diff
@@ -87,30 +87,40 @@ struct TypeErasedParameter {
 
     constexpr size_t to_size() const
     {
-        i64 svalue;
-
-        if (type == TypeErasedParameter::Type::UInt8)
-            svalue = *static_cast<const u8*>(value);
-        else if (type == TypeErasedParameter::Type::UInt16)
-            svalue = *static_cast<const u16*>(value);
-        else if (type == TypeErasedParameter::Type::UInt32)
-            svalue = *static_cast<const u32*>(value);
-        else if (type == TypeErasedParameter::Type::UInt64)
-            svalue = *static_cast<const u64*>(value);
-        else if (type == TypeErasedParameter::Type::Int8)
-            svalue = *static_cast<const i8*>(value);
-        else if (type == TypeErasedParameter::Type::Int16)
-            svalue = *static_cast<const i16*>(value);
-        else if (type == TypeErasedParameter::Type::Int32)
-            svalue = *static_cast<const i32*>(value);
-        else if (type == TypeErasedParameter::Type::Int64)
-            svalue = *static_cast<const i64*>(value);
-        else
+        i64 size_value;
+
+        switch (type) {
+        case Type::UInt8:
+            size_value = static_cast<i64>(*static_cast<u8 const*>(value));
+            break;
+        case Type::UInt16:
+            size_value = static_cast<i64>(*static_cast<u16 const*>(value));
+            break;
+        case Type::UInt32:
+            size_value = static_cast<i64>(*static_cast<u32 const*>(value));
+            break;
+        case Type::UInt64:
+            size_value = static_cast<i64>(*static_cast<u64 const*>(value));
```

Comment:
Isn't it even undefined behavior, due to signed overflow?

Bad suggestion, but still better than UB: `size_t size_value; i64 large_signed;`, all unsigned branches just assign directly to `size_value`, all signed branches do `large_signed = static_cast<>(); VERIFY(large_signed > 0); size_value = large_signed;`
After all, you already have to treat signed and unsigned quite differently.

Reply:
I'm not 100% on the specifics of converting to signed with static_cast.

Reply:
Static_cast here is like "conversion", by rule 8, point 2: https://en.cppreference.com/w/cpp/language/static_cast

Converting to signed with overflow is … wow! Modulo a power of two, and not undefined behavior! https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_conversions

Today I learned something: This code is actually correct, and it's neither UB nor truncation!

Reply:
Can we make it more obviously correct?

Reply:
What about something like: (warning: completely untested)
```c++
    template<typename Visitor>
    constexpr auto visit(Visitor&& visitor) const
    {
        switch (type) {
        case TypeErasedParameter::Type::UInt8:
            return visitor(*static_cast<const u8*>(value));
        case TypeErasedParameter::Type::UInt16:
            return visitor(*static_cast<const u16*>(value));
        case TypeErasedParameter::Type::UInt32:
            return visitor(*static_cast<const u32*>(value));
        case TypeErasedParameter::Type::UInt64:
            return visitor(*static_cast<const u64*>(value));
        case TypeErasedParameter::Type::Int8:
            return visitor(*static_cast<const i8*>(value));
        case TypeErasedParameter::Type::Int16:
            return visitor(*static_cast<const i16*>(value));
        case TypeErasedParameter::Type::Int32:
            return visitor(*static_cast<const i32*>(value));
        case TypeErasedParameter::Type::Int64:
            return visitor(*static_cast<const i64*>(value));
        default:
            TODO();
        }
    }

    constexpr size_t to_size() const
    {
        return visit([](auto value) {
            if constexpr (IsSigned<decltype(value)>)
                VERIFY(value > 0)
            return static_cast<std::size_t>(value);
        });
    }
```

Reply:
Though we also probably want to check that any `u64` / `i64` parameter isn't larger than `NumericLimits<size_t>::max` on 32-bit

Reply:
Using that Visitor pattern, would something like this work?:
```
if constexpr (sizeof<decltype(value)> < sizeof<size_t>)
    VERIFY(value < NumericLimits<size_t>::max)
```

Reply:
Yep, though I think you want to flip the constexpr condition from `<` to `>`
(side note, I have not had enough coffee yet to be mentally parsing all these `<` and `>` :laughing: )

Reply:
Derp, yes, you're right :sweat_smile: 

Fixed my code suggestion to make Andrew's life an epsilon easier.

Reply:
So wait you guys are suggesting this?
```
    constexpr size_t to_size() const 
    {
        return visit([]<typename T>(T value) {
            if constexpr (sizeof(T) > sizeof(size_t))
                VERIFY(value < NumericLimits<size_t>::max());
            if constexpr (IsSigned<T>)
                VERIFY(value > 0);
            return static_cast<size_t>(value);
        });
    }
```
It seems to avoid all the warnings, but do we need both checks? :thinking:

Reply:
Aight well that's what I went with so we can discuss in the next round of review :D

Reply:
I think both are good to have, yes. On x86, a `u64` value of `NumericLimits<size_t>::max() + 1` would become 0 when cast to `size_t`. Probably want to alert the author of that just as much as we'd want to alert them about a negative size.